README.md
# json-stream-parse

Emits JSON objects/arrays from a Buffer stream.

```
const Parser = require("json-stream-parse"),
    net = require("net");

const parser = new Parser(),
    socket = new net.Socket();

parser.on("json", (json) => {
    console.log(json);
});

socket.connect(8888, '127.0.0.1', () => {
    socket.pipe(parser);
})
```

## Events
`json-stream-parser` is a Writeable stream that emits only one event, `json`.  This event is emitted any time a valid JSON object or array is found within the stream.  Any data that is found to be invalid JSON is ignored.

## License
MIT

CHANGELOG.md
# 1.0.3

### BUGFIXES
* Rogue `console.log` statements made their way into the release of `1.0.2`

# 1.0.2

### BUGFIXES
* `algorithm1` would fail to parse messages that were split into multiple packets if one of those packets happened to end in a `}` or `]`.

# 1.0.1
* Changed `algorithm1` to use Buffer.indexOf instead of a manual loop.  This is due to the performance increase with indexOf as it is implemented in C++ rather than raw JS code

test/index.test.js
/* eslint-disable */

const expect = require('chai').expect;

const Parser = require('../lib/index.js');

let parser;

const SOH = 0x01,
    EOT = 0x04;

const BEGIN_MARKER = String.fromCharCode(SOH),
    END_MARKER = String.fromCharCode(EOT);

beforeEach(function() {
    parser = new Parser();
});

describe("_parse", function() {
    it("Should fail to parse", function(done) {
        parser.on("json", (json) => {
            done("Should not have parsed");
        });
        expect(parser._parse("asdf")).to.be.instanceof(SyntaxError);
        done();
    });

    it("Should successfully parse", function(done) {
        parser.on("json", (json) => {
            done();
        });
        expect(parser._parse("{}")).to.be.undefined;
    })
});

describe("setters", function() {
    describe("startMarker", function() {
        it("Should set the start marker with a string", function() {
            parser.startMarker = "a";
            expect(parser.startMarker).to.equal(0x61);
            expect(parser._startMarker).to.equal(0x61);
        });
    
        it("Should fail to set marker with string with more than 1 character", function() {
            let error = null;
            try {
                parseer.startMarker = "asdf";
            } catch (err) {
                error = err;
            }
            expect(error).to.be.instanceof(Error);
            expect(parser.startMarker).to.equal(SOH);
            expect(parser._startMarker).to.equal(SOH);
        });
    
        it("Should set the start marker with a valid number", function() {
            parser.startMarker = 0x00;
            expect(parser.startMarker).to.equal(0x00);
            expect(parser._startMarker).to.equal(0x00);
            parser.startMarker = 0x23;
            expect(parser.startMarker).to.equal(0x23);
            expect(parser._startMarker).to.equal(0x23);
            parser.startMarker = 0x7F;
            expect(parser.startMarker).to.equal(0x7F);
            expect(parser._startMarker).to.equal(0x7F);
        });
    
        it("Should error on invalid input", function() {
            let error = null;
            try {
                parser.startMarker = {};
            } catch (err) {
                error = err;
            }
            expect(error).to.be.instanceof(Error);
            expect(parser.startMarker).to.equal(SOH);
            expect(parser._startMarker).to.equal(SOH);
            error = null;
    
            try {
                parser.startMarker = [];
            } catch (err) {
                error = err;
            }
            expect(error).to.be.instanceof(Error);
            expect(parser.startMarker).to.equal(SOH);
            expect(parser._startMarker).to.equal(SOH);
            error = null;
        });
    
        it("Should error on invalid number range", function() {
            let error = null;
            try {
                parser.startMarker = -1;
            } catch (err) {
                error = err;
            }
            expect(error).to.be.instanceof(Error);
            expect(parser.startMarker).to.equal(SOH);
            expect(parser._startMarker).to.equal(SOH);
            error = null;
    
            try {
                parser.startMarker = 0x80;
            } catch (err) {
                error = err;
            }
            expect(error).to.be.instanceof(Error);
            expect(parser.startMarker).to.equal(SOH);
            expect(parser._startMarker).to.equal(SOH);
            error = null;
        });
    });
    
    describe("endMarker", function() {
        it("Should set the start marker with a string", function() {
            parser.endMarker = "a";
            expect(parser.endMarker).to.equal(0x61);
            expect(parser._endMarker).to.equal(0x61);
        });
    
        it("Should fail to set marker with string with more than 1 character", function() {
            let error = null;
            try {
                parseer.endMarker = "asdf";
            } catch (err) {
                error = err;
            }
            expect(error).to.be.instanceof(Error);
            expect(parser.endMarker).to.equal(EOT);
            expect(parser._endMarker).to.equal(EOT);
        });
    
        it("Should set the start marker with a valid number", function() {
            parser.endMarker = 0x00;
            expect(parser.endMarker).to.equal(0x00);
            expect(parser._endMarker).to.equal(0x00);
            parser.endMarker = 0x23;
            expect(parser.endMarker).to.equal(0x23);
            expect(parser._endMarker).to.equal(0x23);
            parser.endMarker = 0x7F;
            expect(parser.endMarker).to.equal(0x7F);
            expect(parser._endMarker).to.equal(0x7F);
        });
    
        it("Should error on invalid input", function() {
            let error = null;
            try {
                parser.endMarker = {};
            } catch (err) {
                error = err;
            }
            expect(error).to.be.instanceof(Error);
            expect(parser.endMarker).to.equal(EOT);
            expect(parser._endMarker).to.equal(EOT);
            error = null;
    
            try {
                parser.endMarker = [];
            } catch (err) {
                error = err;
            }
            expect(error).to.be.instanceof(Error);
            expect(parser.endMarker).to.equal(EOT);
            expect(parser._endMarker).to.equal(EOT);
            error = null;
        });
    
        it("Should error on invalid number range", function() {
            let error = null;
            try {
                parser.endMarker = -1;
            } catch (err) {
                error = err;
            }
            expect(error).to.be.instanceof(Error);
            expect(parser.endMarker).to.equal(EOT);
            expect(parser._endMarker).to.equal(EOT);
            error = null;
    
            try {
                parser.endMarker = 0x80;
            } catch (err) {
                error = err;
            }
            expect(error).to.be.instanceof(Error);
            expect(parser.endMarker).to.equal(EOT);
            expect(parser._endMarker).to.equal(EOT);
            error = null;
        });
    });
});

describe('_algorithm1', function() {
    let obj = [];

    beforeEach(function() {
        parser.on('json', (json) => {
            obj.push(json);
        });
    });

    afterEach(function() {
        obj = [];
    });

    it('Should skip test when no valid starting character', function(done) {
        parser._buffer = Buffer.from("asdf", 'utf8');

        parser._algorithm1(() => {
            expect(parser._buffer).to.be.instanceof(Buffer);
            expect(parser._buffer.length).length.to.equal(0);
            expect(obj).to.be.empty;
            done();
        });
    });

    it("Should keep buffer if packet is not complete", function(done) {
        parser._buffer = Buffer.from(`[asdf`, 'utf8');

        parser._algorithm1(() => {
            expect(parser._buffer).to.be.instanceof(Buffer);
            expect(parser._buffer).to.deep.equal(Buffer.from(`[asdf`, 'utf8'));
            expect(obj).to.be.empty;
            done();
        });
    });

    it("Should parse one JSON packet", function(done) {
        parser._buffer = Buffer.from(`{}`, 'utf8');

        parser._algorithm1(() => {
            expect(parser._buffer).to.be.instanceof(Buffer);
            expect(parser._buffer.length).length.to.equal(0);
            expect(obj).to.have.lengthOf(1);
            expect(obj[0]).to.deep.eq({});
            done();
        });
    });

    it("Parse two messages in one packet", function(done) {
        parser._buffer = Buffer.from(`{}[]`, 'utf8');

        parser._algorithm1(() => {
            expect(parser._buffer).to.be.instanceof(Buffer);
            expect(parser._buffer.length).length.to.equal(0);
            expect(obj).to.have.lengthOf(2);
            expect(obj[0]).to.deep.eq({});
            expect(obj[1]).to.deep.eq([]);
            done();
        });
    });

    it("Parse one message and buffer the rest", function(done) {
        parser._buffer = Buffer.from(`{}[`, 'utf8');

        parser._algorithm1(() => {
            expect(parser._buffer).to.be.instanceof(Buffer);
            expect(parser._buffer).to.deep.equal(Buffer.from(`[`, 'utf8'));
            expect(obj).to.have.lengthOf(1);
            expect(obj[0]).to.deep.eq({});
            done();
        });
    });

    it("Should parse one message split between two packets", function (done) {
        parser._buffer = Buffer.from(`{"test":"te`, 'utf8');

        parser._algorithm1(() => {
            expect(parser._buffer).to.be.instanceof(Buffer);
            expect(parser._buffer).to.deep.equal(Buffer.from(`{"test":"te`, 'utf8'));
            expect(obj).to.have.lengthOf(0);
            
            parser._buffer = Buffer.concat([
                Buffer.from(`{"test":"te`, 'utf8'),
                Buffer.from(`st"}`, 'utf8')
            ])
            parser._algorithm1(() => {
                expect(parser._buffer).to.be.instanceof(Buffer);
                expect(parser._buffer.length).length.to.equal(0);
                expect(obj).to.have.lengthOf(1);
                expect(obj[0]).to.deep.eq({test:"test"});
                done();
            });
        });
    });

    it("Should parse two packets, one ending with false end", function (done) {
        parser._buffer = Buffer.from(`{"test":{}`, 'utf8');

        parser._algorithm1(() => {
            expect(parser._buffer).to.be.instanceof(Buffer);
            expect(parser._buffer).to.deep.equal(Buffer.from(`{"test":{}`, 'utf8'));
            expect(obj).to.have.lengthOf(0);
            
            parser._buffer = Buffer.concat([
                Buffer.from(`{"test":{}`, 'utf8'),
                Buffer.from(`}`, 'utf8')
            ])
            parser._algorithm1(() => {
                expect(parser._buffer).to.be.instanceof(Buffer);
                expect(parser._buffer.length).length.to.equal(0);
                expect(obj).to.have.lengthOf(1);
                expect(obj[0]).to.deep.eq({test:{}});
                done();
            });
        });
    });

    it("Should parse two messages partially split between two packets", function(done) {
        parser._buffer = Buffer.from(`{"test":"te`, 'utf8');

        parser._algorithm1(() => {
            expect(parser._buffer).to.be.instanceof(Buffer);
            expect(parser._buffer).to.deep.equal(Buffer.from(`{"test":"te`, 'utf8'));
            expect(obj).to.have.lengthOf(0);
            
            parser._buffer = Buffer.concat([
                Buffer.from(`{"test":"te`, 'utf8'),
                Buffer.from(`st"}{}`, 'utf8')
            ])
            parser._algorithm1(() => {
                expect(parser._buffer).to.be.instanceof(Buffer);
                expect(parser._buffer.length).length.to.equal(0);
                expect(obj).to.have.lengthOf(2);
                expect(obj[0]).to.deep.eq({test:"test"});
                expect(obj[1]).to.deep.eq({});
                done();
            });
        });
    });
});

describe('_algorithm2', function() {
    let obj = [];

    beforeEach(function() {
        parser.on('json', (json) => {
            obj.push(json);
        });
    });

    afterEach(function() {
        obj = [];
    });

    it('Should skip test when no valid starting character', function(done) {
        parser._buffer = Buffer.from("asdf", 'utf8');

        parser._algorithm2(() => {
            expect(parser._buffer).to.be.instanceof(Buffer);
            expect(parser._buffer.length).length.to.equal(0);
            expect(obj).to.be.empty;
            done();
        });
    });

    it("Should keep buffer if packet is not complete", function(done) {
        parser._buffer = Buffer.from(`[asdf`, 'utf8');

        parser._algorithm2(() => {
            expect(parser._buffer).to.be.instanceof(Buffer);
            expect(parser._buffer).to.deep.equal(Buffer.from(`[asdf`, 'utf8'));
            expect(obj).to.be.empty;
            done();
        });
    });

    it("Should parse one JSON packet", function(done) {
        parser._buffer = Buffer.from(`{}`, 'utf8');

        parser._algorithm2(() => {
            expect(parser._buffer).to.be.instanceof(Buffer);
            expect(parser._buffer.length).length.to.equal(0);
            expect(obj).to.have.lengthOf(1);
            expect(obj[0]).to.deep.eq({});
            done();
        });
    });

    it("Parse two messages in one packet", function(done) {
        parser._buffer = Buffer.from(`{}[]`, 'utf8');

        parser._algorithm2(() => {
            expect(parser._buffer).to.be.instanceof(Buffer);
            expect(parser._buffer.length).length.to.equal(0);
            expect(obj).to.have.lengthOf(2);
            expect(obj[0]).to.deep.eq({});
            expect(obj[1]).to.deep.eq([]);
            done();
        });
    });

    it("Parse one message and buffer the rest", function(done) {
        parser._buffer = Buffer.from(`{}[`, 'utf8');

        parser._algorithm2(() => {
            expect(parser._buffer).to.be.instanceof(Buffer);
            expect(parser._buffer).to.deep.equal(Buffer.from(`[`, 'utf8'));
            expect(obj).to.have.lengthOf(1);
            expect(obj[0]).to.deep.eq({});
            done();
        });
    });

    it("Should parse one message split between two packets", function (done) {
        parser._buffer = Buffer.from(`{"test":"te`, 'utf8');

        parser._algorithm2(() => {
            expect(parser._buffer).to.be.instanceof(Buffer);
            expect(parser._buffer).to.deep.equal(Buffer.from(`{"test":"te`, 'utf8'));
            expect(obj).to.have.lengthOf(0);
            
            parser._buffer = Buffer.concat([
                Buffer.from(`{"test":"te`, 'utf8'),
                Buffer.from(`st"}`, 'utf8')
            ])
            parser._algorithm2(() => {
                expect(parser._buffer).to.be.instanceof(Buffer);
                expect(parser._buffer.length).length.to.equal(0);
                expect(obj).to.have.lengthOf(1);
                expect(obj[0]).to.deep.eq({test:"test"});
                done();
            });
        });
    });

    it("Should parse two messages partially split between two packets", function(done) {
        parser._buffer = Buffer.from(`{"test":"te`, 'utf8');

        parser._algorithm2(() => {
            expect(parser._buffer).to.be.instanceof(Buffer);
            expect(parser._buffer).to.deep.equal(Buffer.from(`{"test":"te`, 'utf8'));
            expect(obj).to.have.lengthOf(0);
            
            parser._buffer = Buffer.concat([
                Buffer.from(`{"test":"te`, 'utf8'),
                Buffer.from(`st"}{}`, 'utf8')
            ])
            parser._algorithm2(() => {
                expect(parser._buffer).to.be.instanceof(Buffer);
                expect(parser._buffer.length).length.to.equal(0);
                expect(obj).to.have.lengthOf(2);
                expect(obj[0]).to.deep.eq({test:"test"});
                expect(obj[1]).to.deep.eq({});
                done();
            });
        });
    });
});

describe('_algorithm3', function() {
    let obj = [];

    beforeEach(function() {
        parser.on('json', (json) => {
            obj.push(json);
        });
    });

    afterEach(function() {
        obj = [];
    });

    it('Should skip test when no valid starting character', function(done) {
        parser._buffer = Buffer.from("asdf", 'utf8');

        parser._algorithm3(() => {
            expect(parser._buffer).to.be.instanceof(Buffer);
            expect(parser._buffer.length).length.to.equal(0);
            expect(obj).to.be.empty;
            done();
        });
    });

    it("Should keep buffer if packet is not complete", function(done) {
        parser._buffer = Buffer.from(`${BEGIN_MARKER}asdf`, 'utf8');

        parser._algorithm3(() => {
            expect(parser._buffer).to.be.instanceof(Buffer);
            expect(parser._buffer).to.deep.equal(Buffer.from(`${BEGIN_MARKER}asdf`, 'utf8'));
            expect(obj).to.be.empty;
            done();
        });
    });

    it("Should parse one JSON packet", function(done) {
        parser._buffer = Buffer.from(`${BEGIN_MARKER}{}${END_MARKER}`, 'utf8');

        parser._algorithm3(() => {
            expect(parser._buffer).to.be.instanceof(Buffer);
            expect(parser._buffer.length).length.to.equal(0);
            expect(obj).to.have.lengthOf(1);
            expect(obj[0]).to.deep.eq({});
            done();
        });
    });

    it("Parse two messages in one packet", function(done) {
        parser._buffer = Buffer.from(`${BEGIN_MARKER}{}${END_MARKER}${BEGIN_MARKER}[]${END_MARKER}`, 'utf8');

        parser._algorithm3(() => {
            expect(parser._buffer).to.be.instanceof(Buffer);
            expect(parser._buffer.length).length.to.equal(0);
            expect(obj).to.have.lengthOf(2);
            expect(obj[0]).to.deep.eq({});
            expect(obj[1]).to.deep.eq([]);
            done();
        });
    });

    it("Parse one message and buffer the rest", function(done) {
        parser._buffer = Buffer.from(`${BEGIN_MARKER}{}${END_MARKER}${BEGIN_MARKER}[`, 'utf8');

        parser._algorithm3(() => {
            expect(parser._buffer).to.be.instanceof(Buffer);
            expect(parser._buffer).to.deep.equal(Buffer.from(`${BEGIN_MARKER}[`, 'utf8'));
            expect(obj).to.have.lengthOf(1);
            expect(obj[0]).to.deep.eq({});
            done();
        });
    });

    it("Should parse one message split between two packets", function (done) {
        parser._buffer = Buffer.from(`${BEGIN_MARKER}{"test":"te`, 'utf8');

        parser._algorithm3(() => {
            expect(parser._buffer).to.be.instanceof(Buffer);
            expect(parser._buffer).to.deep.equal(Buffer.from(`${BEGIN_MARKER}{"test":"te`, 'utf8'));
            expect(obj).to.have.lengthOf(0);
            
            parser._buffer = Buffer.concat([
                Buffer.from(`${BEGIN_MARKER}{"test":"te`, 'utf8'),
                Buffer.from(`st"}${END_MARKER}`, 'utf8')
            ])
            parser._algorithm3(() => {
                expect(parser._buffer).to.be.instanceof(Buffer);
                expect(parser._buffer.length).length.to.equal(0);
                expect(obj).to.have.lengthOf(1);
                expect(obj[0]).to.deep.eq({test:"test"});
                done();
            });
        });
    });

    it("Should parse two messages partially split between two packets", function(done) {
        parser._buffer = Buffer.from(`${BEGIN_MARKER}{"test":"te`, 'utf8');

        parser._algorithm3(() => {
            expect(parser._buffer).to.be.instanceof(Buffer);
            expect(parser._buffer).to.deep.equal(Buffer.from(`${BEGIN_MARKER}{"test":"te`, 'utf8'));
            expect(obj).to.have.lengthOf(0);
            
            parser._buffer = Buffer.concat([
                Buffer.from(`${BEGIN_MARKER}{"test":"te`, 'utf8'),
                Buffer.from(`st"}${END_MARKER}${BEGIN_MARKER}{}${END_MARKER}`, 'utf8')
            ])
            parser._algorithm3(() => {
                expect(parser._buffer).to.be.instanceof(Buffer);
                expect(parser._buffer.length).length.to.equal(0);
                expect(obj).to.have.lengthOf(2);
                expect(obj[0]).to.deep.eq({test:"test"});
                expect(obj[1]).to.deep.eq({});
                done();
            });
        });
    });
});

gulpfile.js
const { spawn } = require("child_process"),
    eslint = require("gulp-eslint"),
    git = require("gulp-git"),
    gulp = require("gulp"),
    mocha = require("gulp-mocha"),
    os = require("os"),
    projPackage = require("./package"),
    runSequence = require("run-sequence");

/**
 * Run code through a linter.  If there is an error fail
 */
gulp.task("eslint", () => {
    return gulp.src([ "lib/**" ])
        .pipe(eslint())
        .pipe(eslint.format())
        .pipe(eslint.failAfterError());
});

gulp.task("test", () => {
    return gulp.src("./tests/**/*.js", { read: false }).pipe(mocha());
});

gulp.task("tag", (done) => {
    git.tag(projPackage.version, null, (tagErr) => {
        if (tagErr) {
            done(tagErr);
        } else {
            git.push("origin", "master", { args: " --tags" }, (pushErr) => {
                done(pushErr);
            });
        }
    });
});

gulp.task("npm", (done) => {
    let cmd = null;
    if (os.platform() === "win32") {
        cmd = "npm.cmd";
    } else {
        cmd = "npm";
    }
    spawn(cmd, [ "publish" ], { stdio: "inherit" }).on("close", done);
});

gulp.task("publish", (done) => {
    runSequence(
        "eslint",
        "test",
        // "tag",
        "npm",
        done
    );
});


lib/index.js
const EventEmitter = require("eventemitter3"),
    Stream = require("stream");

const Events = {
    Error: "jsonError",
    JSON: "json"
};

const EOT = 0x04,
    SOH = 0x01,
    LSB = 0x5b,
    RSB = 0x5d,
    LCB = 0x7b,
    RCB = 0x7d;

const ALG1 = 1;
const ALG2 = 2;
const ALG3 = 3;

const MIN_ASCII = 0x00;
const MAX_ASCII = 0x7F;

const DEF_ALG = ALG1;

class Parser extends Stream.Writable {

    constructor () {
        super();

        this._buffer = Buffer.alloc(0);
        this._emitter = new EventEmitter();
        this._searchIndex = 0;
        this._startIndex = -1;
        this._numBrackets = 0;
        this._startMarker = SOH;
        this._endMarker = EOT;
        this._algorithm = DEF_ALG;
    }

    get algorithm () {
        return this._algorithm;
    }

    get startMarker () {
        return this._startMarker;
    }

    get endMarker () {
        return this._endMarker;
    }

    set algorithm (value) {
        this._algorithm = value;
    }

    set startMarker (value) {
        if (typeof value === "string" && value.length === 1) {
            this._startMarker = value.charCodeAt(0);
        } else if (typeof value === "number" &&
            value >= MIN_ASCII &&
            value <= MAX_ASCII) {

            this._startMarker = value;
        } else {
            throw new Error("Invalid start marker");
        }
    }

    set endMarker (value) {
        if (typeof value === "string") {
            this._endMarker = value.charCodeAt(0);
        } else if (typeof value === "number" &&
            value >= MIN_ASCII &&
            value <= MAX_ASCII) {

            this._endMarker = value;
        } else {
            throw new Error("Invalid start marker");
        }
    }

    on (event, cb) {
        switch (event) {
        case Events.JSON:
            this._emitter.on(Events.JSON, cb);
            break;
        case Events.Error:
            this._emitter.on(Events.Error, cb);
            break;
        default:
            super.on(event, cb);
        }
    }

    _parse (str) {
        let obj = null;
        try {
            obj = JSON.parse(str);
        } catch (error) {
            if (error instanceof SyntaxError) {
                return error;
            }

            throw error;
        }

        this._emitter.emit(Events.JSON, obj);
    }

    /**
     * Parser A looks for repeated JSON objects (i.e. "][", "}{", "}[",
     * or "]{").
     *
     * @param {Function} done - Function to call when the current buffer is
     *                          fully parsed
     * @returns {undefined}
     */
    _algorithm1 (done) {
        let endChar = null;

        this._startChar = this._buffer[0];
        if (this._buffer[0] === LSB) {
            endChar = RSB;
        } else if (this._buffer[0] === LCB) {
            endChar = RCB;
        } else {
            this._buffer = Buffer.alloc(0);
            return done();
        }

        const index = this._getEndIndex(endChar);

        // Packet isn't complete
        if (index === -1 && this._buffer[this._buffer.length - 1] !== endChar) {
            this._searchIndex = this._buffer.length;
            done();
        } else if (index === -1) {
            // Packet might be complete
            if (this._parse(this._buffer.toString("utf8"))) {
                this._searchIndex = this._buffer.length;
            } else {
                this._buffer = Buffer.alloc(0);
                this._searchIndex = 0;
            }
            done();
        } else {
            this._parse(this._buffer.toString("utf8", 0, index + 1));
            this._buffer = this._buffer.slice(index + 1);
            this._searchIndex = 0;
            this._algorithm1(done);
        }
    }

    _getEndIndex (char) {
        const indexA = this._buffer.indexOf(
            `${String.fromCharCode(char)}[`,
            this._searchIndex
        );
        const indexB = this._buffer.indexOf(
            `${String.fromCharCode(char)}{`,
            this._searchIndex
        );
        let index = null;

        if (indexA === -1) {
            index = indexB;
        } else if (indexB === -1) {
            index = indexA;
        } else {
            index = indexA < indexB ?
                indexA :
                indexB;
        }

        return index;
    }

    /**
     * AlgorithmB keeps track of JSON brackets to tell when a full JSON
     * object/array has been recevied.
     *
     * @param {Function} done Function to call when done parsing
     * @returns {undefined}
     */
    _algorithm2 (done) {
        // Get starting bracket
        if (this._numBrackets === 0) {
            for (let i = 0; i < this._buffer.length; i += 1) {
                if (this._buffer[i] === LSB) {
                    this._startB = LSB;
                    this._endB = RSB;
                    this._startIndex = i;
                    this._searchIndex = i + 1;
                    this._numBrackets = 1;
                    break;
                } else if (this._buffer[i] === LCB) {
                    this._startB = LCB;
                    this._endB = RCB;
                    this._startIndex = i;
                    this._searchIndex = i + 1;
                    this._numBrackets = 1;
                    break;
                }
            }
        }

        if (this._startIndex >= 0) {
            for (let i = this._searchIndex; i < this._buffer.length; i += 1) {
                if (this._buffer[i] === this._startB) {
                    this._numBrackets += 1;
                } else if (this._buffer[i] === this._endB) {
                    this._numBrackets -= 1;
                }

                if (this._numBrackets === 0) {
                    this._parse(this._buffer.slice(this._startIndex, i + 1));
                    this._buffer = this._buffer.slice(i + 1);
                    return this._algorithm2(done);
                }
            }
        } else {
            this._buffer = Buffer.alloc(0);
        }

        if (this._buffer.length > 0) {
            this._searchIndex = this._buffer.length;
        }
        done();
    }

    /**
     * AlgorithmE uses start/stop characters to know when a JSON structure has
     * started or ended.  This also uses a Buffer instead of a String input
     *
     * @param {Function} done Function to call when done parsing
     * @returns {undefined}
     */
    _algorithm3 (done) {
        let end = null;

        // If the start index is set then we found it on a previous iteration
        if (this._startIndex < 0) {
            this._startIndex = this._buffer.indexOf(this._startMarker);
        }

        if (this._startIndex >= 0) {
            end = this._buffer.indexOf(this._endMarker, this._searchIndex);
        } else {
            this._buffer = Buffer.alloc(0);
            return done();
        }

        if (end >= 0) {
            this._parse(this._buffer.toString(
                "utf8",
                this._startIndex + 1,
                end
            ));
            this._buffer = this._buffer.slice(end + 1);
            this._startIndex = -1;
            this._searchIndex = 0;
            this._algorithm3(done);
        } else {
            this._buffer = this._buffer.slice(this._startIndex);
            this._searchIndex = this._buffer.length;
            done();
        }
    }

    // ///////////////// //
    // Transform Methods //
    // ///////////////// //

    _write (chunk, encoding, cb) {
        this._buffer = Buffer.concat([
            this._buffer,
            chunk
        ]);

        switch (this._algorithm) {
        case ALG1:
            this._algorithm1(cb);
            break;
        case ALG2:
            this._algorithm2(cb);
            break;
        case ALG3:
            this._algorithm3(cb);
            break;
        default:
            throw new Error("Inavlid parser");
        }
    }

    // ///////////////////// //
    // End Transform Methods //
    // ///////////////////// //

}

module.exports = Parser;
